tests/testthat/test-derive_date_vars.R

```diff
@@ -0,0 +1,895 @@
+    "2003-12-15T13:59:59",
+    "2020-07-31T23:59:59"
+  )
+  expect_equal(
+    impute_dtc_dtm(
+      dtc = input,
+      highest_imputation = "D",
+      date_imputation = "LAST",
+      time_imputation = "LAST",
+      preserve = FALSE
+    ),
+    expected_output
+  )
+})
+
+## Test 5: impute date to last, time to 23:59:59 and preserve = TRUE ----
+test_that("impute_dtc_dtm Test 5: impute date to last, time to 23:59:59 and preserve = TRUE", {
+  expected_output <- c(
+    "2019-07-18T15:25:40.243",
+    "2019-07-18T15:25:40",
+    "2019-07-18T15:25:59",
+    "2019-07-18T23:59:59",
+    "2019-02-28T23:59:59",
+    "2019-12-31T23:59:59",
+    "2019-12-07T23:59:59",
+    "2003-12-15T23:15:18",
+    "2003-12-15T13:59:19",
+    "2020-07-31T00:00:59"
+  )
+  expect_equal(
+    imputes <- impute_dtc_dtm(
+      dtc = input,
+      highest_imputation = "M",
+      date_imputation = "LAST",
+      time_imputation = "LAST",
+      preserve = TRUE
+    ),
+    expected_output
+  )
+})
+
+## Test 6: no date imputation, impute second to 59 ----
+test_that("impute_dtc_dtm Test 6: no date imputation, impute second to 59", {
+  expected_output <- c(
+    "2019-07-18T15:25:40.243",
+    "2019-07-18T15:25:40",
+    "2019-07-18T15:25:59",
+    NA_character_,
+    NA_character_,
+    NA_character_,
+    NA_character_,
+    NA_character_,
+    NA_character_,
+    NA_character_
+  )
+
+  expect_equal(
+    imputes <- impute_dtc_dtm(
+      dtc = input,
+      highest_imputation = "s",
+      time_imputation = "LAST",
+      preserve = FALSE
+    ),
+    expected_output
+  )
+})
+
+## Test 7: impute date to mid, time to last ----
+test_that("impute_dtc_dtm Test 7: impute date to mid, time to last", {
+  expected_output <- c(
+    "2019-07-18T15:25:40.243",
+    "2019-07-18T15:25:40",
+    "2019-07-18T15:25:00",
+    "2019-07-18T00:00:00",
+    "2019-02-15T00:00:00",
+    "2019-06-30T00:00:00",
+    "2019-06-30T00:00:00",
+    "2003-12-15T00:00:00",
+    "2003-12-15T13:00:00",
+    "2020-07-15T00:00:00"
+  )
+  expect_equal(
+    imputes <- impute_dtc_dtm(
+      dtc = input,
+      highest_imputation = "M",
+      date_imputation = "MID",
+      time_imputation = "FIRST",
+      preserve = FALSE
+    ),
+    expected_output
+  )
+})
+
+## Test 8: min_dates parameter works ----
+test_that("impute_dtc_dtm Test 8: min_dates parameter works", {
+  expect_equal(
+    impute_dtc_dtm(c("2020-12", "2020-11", NA_character_),
+      min_dates = list(
+        c(
+          ymd_hms("2020-12-06T12:12:12"),
+          NA,
+          NA
+        ),
+        c(
+          ymd_hms("2020-11-11T11:11:11"),
+          ymd_hms("2020-11-11T11:11:11"),
+          ymd_hms("2020-11-11T11:11:11")
+        )
+      ),
+      highest_imputation = "Y",
+      date_imputation = "first"
+    ),
+    c("2020-12-06T12:12:12", "2020-11-11T11:11:11", "2020-11-11T11:11:11")
+  )
+})
+
+## Test 9: max_dates parameter works ----
+test_that("impute_dtc_dtm Test 9: max_dates parameter works", {
```

Comment:
you could add a test case for the following that didn't work before just to be sure:
`impute_dtc(dtc = "2020-02-02", time_imputation = "last", max_dates = list(lubridate::ymd("2020-02-02")))` gives `"2020-02-02T00:00:00"`, but should really be `"2020-02-02T23:59:59"`

Reply:
Test case extended